### ISSUE TITLE
Add FTP server as alternative for fuse

### ISSUE THREAD

**Issue by askeron** [restic/restic#485]
Original issue title: Add WebDAV server as alternative for fuse

Important: This is not about an WebDAV Backend.

I think the idea of using fuse to examine the backuped snapshots is great, but as some OS do not support fuse and somebody might not have enough rights to use fuse, it would be great to have an alternative.
My idea was to add a command to restic that instead of mounting the snapshots via fuse opens a local WebDAV server. As all major OS have a options to mount a WebDAV server, everybody would be able to mount the snapshots.

As I guess that it is easy to find a WebDAV serverlib for Go and you would use the same directory structure the fuse mount, it should not be too complicated to add this feature.

**Comment by fd0** [comment]
Good idea, although we need to investigate if it's really necessary to implement WebDAV, as opposed to e.g. ftp or plain HTTP.

**Comment by askeron** [comment]
Thanks. I also thought about HTTP and FTP, but the advantage of WebDAV would be, that you can mount it easily.

**Comment by fd0** [comment]
You can easily mount an ftp server on a recent Windows: http://www.thewindowsclub.com/map-an-ftp-drive-windows

**Comment by askeron** [comment]
Oh cool, I did not know that. Then a FTP server is the better solution.

**Comment by TobiX** [comment]
Just my humble €0.02: Please don't use FTP. It's not a simple protocol. It's an old, legacy protocol that [needs to die](http://mywiki.wooledge.org/FtpMustDie). Even if this is only for localhost connections - FTP doesn't even define how directory listings should look. And look, there is a nice WebDAV server library for Go right here: https://godoc.org/golang.org/x/net/webdav - Using WebDAV instead of plain HTTP gives instant integration into most file managers: Windows Explorer, Apple Finder, KDE, GNOME.

**Comment by fd0** [comment]
interesting, thanks for the comment

**Comment by fd0** [comment]
argh, wrong button, sorry.

**Comment by stv0g** [comment]
I don't get the point.
Which major OS doesn't support FUSE?

BSDs, OS X, Linux, Windows... all of them have a FUSE driver..

(Windows for example https://github.com/dokan-dev/dokany)

**Comment by fd0** [comment]
Dokan is not the same as FUSE, it's a compatibility layer around libfuse (which we do not use), I've explained it here: https://github.com/restic/restic/issues/361#issuecomment-192162082

**Comment by wrouesnel** [comment]
WebDAV on Windows is finicky but it does work (in a way which is compatible with everyone else too).  More important is that WebDAV gives a super-easy way to allow users to restore single files from snapshots with low setup overhead (i.e. `rustic web --port :8080` and then collect the repository password per-session on a local network).

**Comment by aliron19** [comment]
Please, take into account that WSL requires Hyper-V to work, but it cannot coexist with any other hypervisor software the user may already be using (like VirtualBox, etc). So, if someone is already using VirtualBox (or any another hypervisor), they will not be able to use WSL.

In summary, I think that WSL should not be taken for granted for every Windows user.

**Comment by hi2u** [comment]
It might be worth looking at how Kopia does its mounting on Windows.  Looks like it does use webdav.

**Comment by fd0** [comment]
Having a second way to access the contents of the repo in a way similar to a file system becomes more important as the [source code for osxfuse is [not available any more](https://github.com/osxfuse/osxfuse/issues/590) and the library we're using for fuse [dropped support for macOS entirely](https://github.com/bazil/fuse/issues/224).

**Comment by technicalguy** [comment]
I'm super keen for this – is there a way I can contribute monetarily and/or might someone be able to give me some mentoring for me to attempt adding this feature? :smile:

**Comment by lucasew** [comment]
Well. Fuse works but in Windows it requires winfsp, that requires admin to install. Also there is the MacOS question that the library dropped support and the limitations of using WSL2 to workaround this. 

Just for viewing stuff, WebDAV, or even plain HTTP, can solve this issue for people that are using Windows or MacOS. Maybe a route in the restic server?

**Comment by askeron** [comment]
I'd would really, really like to get this feature. It should rather be simple to somebody who knows Go, as there are webdav-server-implementation as libraries and there is already the mount command. It should nearly as simple as connecting these two code parts and write a command for it. I even tried to learn go to build this myself, but I did not get used to it.

Here are some things I could offer to give an incentive to somebody to write this feature: I would be willing to give 100 EUR to someone who builds this feature, or instead as I am a senior developer in Java and Kotlin and my NodeJS and Docker skills are also not bad, I would offer an developer exchange, so in exchange I would work twice the time you needed for this issue on your project, or the project of your choise.

**Comment by wrouesnel** [comment]
I've banged together and locally tested another prototype of adding WebDAV support to the restic (as `restic serve`) as a fun exercise today. There's already another PR up which takes a similar approach but it looks like it hasn't made any progress for a while: https://github.com/wrouesnel/restic/tree/wrouesnel/add_support_for_webdav

**Comment by ducalex** [comment]
> https://github.com/wrouesnel/restic/tree/wrouesnel/add_support_for_webdav

I found a very old PR from you but none for this new prototype, is there any reason why you (seemingly) abandoned it? Because your new code seems straightforward and more importantly self-contained, so it has a decent chance of being accepted.

I have a pull request open (#4276) that allows to browse repositories in a web browser but there hasn't been much interest from maintainers. So I'd be interested in maybe finishing your work because perhaps WebDAV would gather more interest!

**Comment by wrouesnel** [comment]
@ducalex It's open source so go for it! My use-case for WebDAV. I ran the build locally for a while but the system I was using it with no longer exists.
